Feature/dashboard & about us

elornHamon opened the pull request:
-  Add Dashboard view 
-  Add Dismissible AboutUs Card to Dashboard view (Horizontal swipe to dismiss card, top right restore button to restore card)

![image](https://user-images.githubusercontent.com/66817481/118056101-4a006a80-b357-11eb-8ea0-ccfe1a1cf9d1.png)
![image](https://user-images.githubusercontent.com/66817481/118056118-55ec2c80-b357-11eb-8d40-dc5d97c6bafe.png)

AlexandreMarkus commented on lib/ui/views/dashboard_view.dart:
```diff
@@ -0,0 +1,154 @@
+              const SizedBox(width: 10),
+              TextButton(
+                onPressed: () {
+                  launch('https://github.com/ApplETS/Notre-Dame');
```
use the strings in urls.dart

AlexandreMarkus commented on lib/ui/views/dashboard_view.dart:
```diff
@@ -0,0 +1,154 @@
+              const SizedBox(width: 10),
+              TextButton(
+                onPressed: () {
+                  launch('mailto:info@clubapplets.ca');
```
add this string to the file urls.dart

AlexandreMarkus commented on lib/ui/views/dashboard_view.dart:
```diff
@@ -0,0 +1,154 @@
+            title: Text(AppIntl.of(context).title_dashboard),
+            centerTitle: false,
+            automaticallyImplyLeading: false,
+            actions: _buildActionButtons(),
```
add the widget here

elornHamon replied:
Not sure that I get this one, could you elaborate ? 😕

AlexandreMarkus replied:
Copy lines 136-141 here instead of calling the function

AlexandreMarkus commented on lib/ui/views/dashboard_view.dart:
```diff
@@ -0,0 +1,154 @@
+            automaticallyImplyLeading: false,
+            actions: _buildActionButtons(),
+          ),
+          body: Column(mainAxisSize: MainAxisSize.min, children: [
```
use a listview instead to prevent this to happen. Maybe a listview.builder could work also
![image](https://user-images.githubusercontent.com/59664238/111038838-f0cc9700-83f8-11eb-8ea0-47b63a4dcc61.png)

AlexandreMarkus commented on test/ui/views/dashboard_view_test.dart:
```diff
@@ -0,0 +1,95 @@
+        expect(find.text("EMAIL"), findsOneWidget);
+      });
+
+      testWidgets('AboutUsCard is dismissible and can be restored',
```
Good test !! :smile:

MysticFragilist commented on lib/core/viewmodels/dashboard_viewmodel.dart:
```diff
@@ -0,0 +1,24 @@
+// FLUTTER / DART / THIRD-PARTIES
+import 'package:flutter/cupertino.dart';
```
It's more a question for anyone who knows, but I've seen the cupertino and material import, it can be used to import the `@required` in ctor, is there any difference between those two imports?

elornHamon replied:
I'm not sure about the difference either, it seems https://flutter.dev/docs/development/ui/widgets/cupertino is the IOs style widget library but maybe @AlexandreMarkus knows more about that ? I'll be replacing this with the material.dart import as cupertino imports seem to have been removed from other widgets as well.

apomalyn commented on lib/core/managers/dashboard_manager.dart:
```diff
@@ -0,0 +1,28 @@
+// OTHER
+import 'package:notredame/locator.dart';
+
+class DashboardManager with ChangeNotifier {
```
I'm sure to understand the purpose of this manager :thinking:. We could simply call the PreferencesService directly in the view model. What do you think about it?

apomalyn reviewed (changes requested):
The padding-top (and so bottom) are a little too thin I think :thinking:

MysticFragilist commented:
Hi! I tried to generate the golden file, but I was faced with this error. I don't know if anyone has had this error when updating the golden before?
![Screenshot from 2021-03-29 21-20-25](https://user-images.githubusercontent.com/25663435/112919151-7c6e4500-90d4-11eb-98ed-07d95a6172f1.png)
The code that overflow is the following (dashboard_view.dart:97):
```dart
Row(children: [
  const SizedBox(width: 10),
  TextButton(
    onPressed: () {
      Utils.launchURL(Urls.clubFacebook, AppIntl.of(context));
    },
    child: Text(AppIntl.of(context).facebook.toUpperCase(),
        style: Theme.of(context).primaryTextTheme.button),
  ),
  const SizedBox(width: 10),
  TextButton(
    onPressed: () {
      Utils.launchURL(Urls.clubGithub, AppIntl.of(context));
    },
    child: Text(AppIntl.of(context).github.toUpperCase(),
        style: Theme.of(context).primaryTextTheme.button),
  ),
  const SizedBox(width: 10),
  TextButton(
    onPressed: () {
      Utils.launchURL(Urls.clubEmail, AppIntl.of(context));
    },
    child: Text(AppIntl.of(context).email.toUpperCase(),
        style: Theme.of(context).primaryTextTheme.button),
  ),
]),
```
@elornHamon my guess is that you will need to use padding element instead of fixed SizedBox widget. But perhaps it's something else...
Here is the golden file generated (you can see the yellow/black band that overflowed)
![dashboardView_1](https://user-images.githubusercontent.com/25663435/112919145-78dabe00-90d4-11eb-85f5-c33340d59c47.png)

AlexandreMarkus commented:
> Hi! I tried to generate the golden file, but I was faced with this error. I don't know if anyone has had this error when updating the golden before?
> ![Screenshot from 2021-03-29 21-20-25](https://user-images.githubusercontent.com/25663435/112919151-7c6e4500-90d4-11eb-98ed-07d95a6172f1.png)
> The code that overflow is the following (dashboard_view.dart:97):
> 
> ```dart
> Row(children: [
>   const SizedBox(width: 10),
>   TextButton(
>     onPressed: () {
>       Utils.launchURL(Urls.clubFacebook, AppIntl.of(context));
>     },
>     child: Text(AppIntl.of(context).facebook.toUpperCase(),
>         style: Theme.of(context).primaryTextTheme.button),
>   ),
>   const SizedBox(width: 10),
>   TextButton(
>     onPressed: () {
>       Utils.launchURL(Urls.clubGithub, AppIntl.of(context));
>     },
>     child: Text(AppIntl.of(context).github.toUpperCase(),
>         style: Theme.of(context).primaryTextTheme.button),
>   ),
>   const SizedBox(width: 10),
>   TextButton(
>     onPressed: () {
>       Utils.launchURL(Urls.clubEmail, AppIntl.of(context));
>     },
>     child: Text(AppIntl.of(context).email.toUpperCase(),
>         style: Theme.of(context).primaryTextTheme.button),
>   ),
> ]),
> ```
> 
> @elornHamon my guess is that you will need to use padding element instead of fixed SizedBox widget. But perhaps it's something else...
> Here is the golden file generated (you can see the yellow/black band that overflowed)
> ![dashboardView_1](https://user-images.githubusercontent.com/25663435/112919145-78dabe00-90d4-11eb-85f5-c33340d59c47.png)

Yes, I had the same issue. To fix this I told @elornHamon to use Wrap instead of the Row

camillebrulotte commented:
Unless "PreferencesFlag.scheduleCard" and "PreferencesFlag.progressBarCard" is not normal, everything seems to work fine. I
tested it on IOS and Android. :)

<img src="https://user-images.githubusercontent.com/37625944/116472642-2a8c1c80-a844-11eb-8617-a8e278db69c4.png" height="600">

elornHamon commented:
> Unless "PreferencesFlag.scheduleCard" and "PreferencesFlag.progressBarCard" is not normal, everything seems to work fine. I
> tested it on IOS and Android. :)
> 
> <img alt="" height="600" src="https://user-images.githubusercontent.com/37625944/116472642-2a8c1c80-a844-11eb-8617-a8e278db69c4.png">

Thanks for the review 😄 Yup, they are just there to prove that reordering/dismissing works for now. They'll be progressively replaced by the real cards when we merge them.

AlexandreMarkus commented on lib/ui/router.dart:
```diff
@@ -3,6 +3,7 @@ import 'package:flutter/material.dart';
 
 // ROUTES
 import 'package:notredame/core/constants/router_paths.dart';
+import 'package:notredame/ui/views/dashboard_view.dart';
```
Same

AlexandreMarkus commented on lib/ui/views/dashboard_view.dart:
```diff
@@ -0,0 +1,184 @@
+// VIEWMODEL
+import 'package:notredame/core/viewmodels/dashboard_viewmodel.dart';
+
+// WIDGET
```
```suggestion
// WIDGETS
```

AlexandreMarkus reviewed (commented):
Nice job! Pumped up to see the other cards!

AlexandreMarkus commented:
@elornHamon  @apomalyn  Is our goal is to close this issue then add the other cards in other issues?

elornHamon commented:
> @elornHamon @apomalyn Is our goal is to close this issue then add the other cards in other issues?

Yes that seems to be the idea in order to keep the PRs small. I still need to push a few code quality refactors and some more tests requested by @apomalyn and this one should be good. The next cards PRs should follow promptly.

apomalyn commented on test/viewmodels/dashboard_viewmodel_test.dart:
```diff
@@ -0,0 +1,178 @@
+      tearDownFlutterToastMock();
+    });
+
+    group("futureToRun - ", () {
```
You could add a test to see what happens if the settings manager behave correctly

elornHamon replied:
Do you mean incorrectly ? Or is there more to test when it behaves correctly ?

apomalyn replied:
You right I wanted to say 'incorrectly' 😅

apomalyn reviewed (changes requested):
Some littles changes ^^

elornHamon commented:
> So I don't have any issues with the code, but there is some test missing 😅
> 
> * SettingsManager -> getDashboard()
> * DashboardViewModel -> setCardVisible(..)
> * There is a problem with the test: 'AboutUsCard is reorderable and can be restored', it seems like the reorder function is never called. Can you investigate it?
> 
> Otherwise, it seems everything is good 😄
> 
> Reminder: to see the coverage add `--coverage` to the test command then at the root of the project use `genhtml coverage/lcov.info -o coverage/html`

I added the missing tests for the Settings manager. For the view the onReorder and onDismissed functions don't appear as covered in the coverage report for some reason, I don't really know how to deal with it as the cards are definitely dismissed and reordered in the tests though, which should trigger the call. I also removed setCardVisible as it was never used and would have needed a bit more work on the logic in order for it to work properly. I'll add it again (with proper tests) if we end up having a use case for it.

apomalyn commented:
@elornHamon I will investigate why ^^'

camillebrulotte commented on lib/ui/views/dashboard_view.dart:
```diff
@@ -0,0 +1,181 @@
+import 'package:notredame/ui/utils/loading.dart';
+
+// OTHER
+import 'package:notredame/ui/utils/app_theme.dart';
```
Il pourrait aller avec les autres imports de utils ! :D

elornHamon replied:
Bien vu ! 😃

elornHamon commented:
> @elornHamon I will investigate why ^^'

Thanks to you input I managed to fix the behaviour of the test and the onReorder function is properly covered now 🎉 : 
![image](https://user-images.githubusercontent.com/66817481/117582990-ce8e8700-b0d2-11eb-9305-cb049d786b79.png)

apomalyn reviewed (approved):
LGTM 🎉 👍 🍾

AlexandreMarkus commented on lib/ui/widgets/dismissible_card.dart:
```diff
@@ -0,0 +1,30 @@
+      key: UniqueKey(),
+      onDismissed: onDismissed,
+      child: Card(
+        elevation: elevation,
```
```suggestion
        elevation: elevation,
        margin: const EdgeInsets.all(0.0),
```

AlexandreMarkus replied:
this will remove the white rectangle around the card when you long press to reorder

elornHamon replied:
Woah good call 👍 , is there a default margin non-zero margin applied to the card widget ?

AlexandreMarkus commented on lib/ui/views/dashboard_view.dart:
```diff
@@ -0,0 +1,179 @@
+                  ? buildLoading()
+                  : ReorderableListView(
+                      onReorder: (oldIndex, newIndex) =>
+                          onReorder(model, oldIndex, newIndex),
```
```suggestion
                          onReorder(model, oldIndex, newIndex),
                          padding: const EdgeInsets.all(8.0),
```

AlexandreMarkus replied:
this will add the padding around all cards

AlexandreMarkus commented:
immediate approval after the 2 little changes above

elornHamon commented:
> immediate approval after the 2 little changes above @elornHamon
> Also make sure to be updated with master

Changes applied, could you please regenerate the golden files on linux when you have a minute ?

AlexandreMarkus commented:
> > immediate approval after the 2 little changes above @elornHamon
> > Also make sure to be updated with master
> 
> Changes applied, could you please regenerate the golden files on linux when you have a minute ?

Done but there is one test that is not passing anymore

AlexandreMarkus reviewed (approved):
Close to the end !!!!!

AlexandreMarkus commented:
@elornHamon it is in your hands now!!!
